### PR TITLE
 chore: protect organization endpoints with license

### LINKED ISSUE #13969: Protect org CRUD endpoints with multi-organization experiment
Many of the existing organization endpoints existed before the experiment existed. Because of that, the endpoints are not protected and can be used by any AGPL endpoint.

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

Closes https://github.com/coder/coder/issues/13969

Creating, editing, and deleting an organization now require a license (and experiment). Endpoints are in the `/enterprise` directory now.

Any tests that created an organization in the AGPL directory didn't make sense anyway. They had no means of having a provisioner daemon, so had no functionality besides adding/removing members.

**This PR is 99% moving tests from AGPL to Enterprise**. And changing 
```golang
client := coderdtest.New(t, nil)
coderdtest.CreateFirstUser(t, client)
```

To 

```golang
dv := coderdtest.DeploymentValues(t)
dv.Experiments = []string{string(codersdk.ExperimentMultiOrganization)}
client, user := coderdenttest.New(t, &coderdenttest.Options{
	Options: &coderdtest.Options{
		DeploymentValues: dv,
	},
	LicenseOptions: &coderdenttest.LicenseOptions{
		Features: license.Features{
			codersdk.FeatureMultipleOrganizations: 1,
		},
	},
})
```

### Test Coverage

There is a chance our test coverage will go down since we probably do not track cross pkg test coverage. To try and maintain the coverage and move the tests sounds like a massive headache. We can refactor some of these tests and bring them back into AGPL if it is an issue. Refactoring all the tests now would not be ideal since we need multi-org tests.

I am sure we have redundant tests, but there is no easy way to identify them.

<details>
  <summary>Ignore, text before I found a good solution</summary>

Enterprise has an issue overriding sub endpoints. The only way to properly extend them is to save them somewhere, and reference the sub routers directly. 

When trying to "merge" routes, you either overwrite the endpoint, or get a panic.

https://goplay.tools/snippet/6JXkSIdJ997 

-----

# Alternatives

## Duplicate routes

One idea is to just replace the entire sub router in Enterprise. Essentially copying the routes to both places, and using a unit test to ensure Enterprise is a super set.

This unfortunately cannot work because you cannot overwrite a route in chi.
```
panic: chi: attempting to Mount() a handler on an existing path, '/organizations'
```

## Write some `r.Merge(r1, r2)` or `r.Extend("/organizations")` function

The idea is to fetch the router dynamically, and extend it. Rather than hard coding it in some field in the `API` struct. Unfortunately traversing the router is only supported by `r.Routes()` and `chi.Walk()`. Both expose the `http.Handler` but **not** the `router`. So you cannot fetch the `chi.Mux` dynamically without some serious changes :cry: 

## Write a `MergeRouters(a, b chi.Routes)`

This is possible, but such a pain. Essentially you can call `r.Routes()` to traverse all the sub routers and endpoints. If you assume no typecasting (`route.(*chi.Mux)`), then you cannot extend these routes, only read them.

So to merge them, we'd have to build our own `/pattern` tree (handling things like `r.Route("/nested/stuff")` being depth=2 in the tree.

Then merge 2 route trees that we build.

Then convert that route tree back to a chi.Mux by doing a depth first search and doing the
 ```golang
chi.Route("/pattern", func(r chi.Router) {
  // When you do `r.Group` it creates a fresh middleware stack.
  // I'm not sure how this looks in the `.Routes()` response. Our tree
  // would have to be able to have different middleware stacks for the
  // same pattern nodes.
  r.Use(treeNode.Middlewares...)

  for _, subRouter := range treeNode.Routers {
    dfs(r, subRouter)
  }
  for _, endpoint := range treeNode.Endpoints {
    r.Method(endpoint.Method, endpoint.Pattern, endpoint.Handler)
  }
})
```

This really feels like we'd just writing our own `http.Mux`. It can be done, and it's _not that challenging_, it's just a lot of leg work that I don't find very useful at this time.
  
</details>

### CODE REVIEW

File: coderd/coderd.go [hunk 1]
```diff
@@ -1336,6 +1335,13 @@ type API struct {
 
 	// APIHandler serves "/api/v2"
 	APIHandler chi.Router
+	// APISubRoutes are AGPL sub routers that are saved for enterprise
+	// to add endpoints to. Chi does not allow conflicting routers to
+	// exist, so this is the easier way to "merge" routers.
+	APISubRoutes struct {
+		Organizations      chi.Router
+		SingleOrganization chi.Router
+	}
```
Comment (Emyrk) [inline coderd/coderd.go:1344]: This is where I expose some sub-routers

File: enterprise/coderd/coderd.go [hunk 2]
```diff
@@ -181,6 +181,24 @@ func New(ctx context.Context, options *Options) (_ *API, err error) {
...
+
+	api.AGPL.APISubRoutes.SingleOrganization.Group(func(r chi.Router) {
+		r.Use(
+			api.RequireFeatureMW(codersdk.FeatureMultipleOrganizations),
+			httpmw.RequireExperiment(api.AGPL.Experiments, codersdk.ExperimentMultiOrganization),
+		)
+		r.Patch("/", api.patchOrganization)
+		r.Delete("/", api.deleteOrganization)
+	})
+
```
Comment (Emyrk) [inline enterprise/coderd/coderd.go:201]: And where I add the enterprise routes to the AGPL router saved in the struct.

**Review by f0ssel — APPROVED** [review]
Glad we are doing this. The fact that none of the golden files, API docs, or codersdk code changed with all of this gives me confidence this should be pretty safe.

**Comment by Emyrk** [comment]
Rebasing in case any new tests were added that I need to move

**Pull request merged**